### PR TITLE
texmaker 5.1.2 (new formula)

### PR DESCRIPTION
**Pull request by albertz**
https://www.xm1math.net/texmaker/

Esp I wanted to have some build with Qt 6 on Linux, as the Qt 5 PDF viewer does not handle my High DPI monitor correctly (looks pixelated). This seems to be fixed in Qt 6.

### CODE REVIEW

**Comment by albertz** [comment]
Note that I reported the patch already upstream. But I guess it needs a new version. There is unfortunately no public repository for the project.

File: Formula/texmaker.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  license "GPL-2.0-only"
+
+  depends_on "qt"
+
+  on_linux do
+    depends_on "gcc"
+  end
+  on_macos do
+    depends_on xcode: :build
+  end
```
Comment (carlocab) [inline Formula/texmaker.rb:15]: `depends_on :xcode` is a no-op for Linux, so you don't need to wrap this in an `on_macos` block.

File: Formula/texmaker.rb [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  desc "Free cross-platform LaTeX editor since 2003"
+  homepage "https://www.xm1math.net/texmaker/"
+  url "https://www.xm1math.net/texmaker/texmaker-5.1.2.tar.bz2"
+  sha256 "526896f2c1ae561130eec7aae815b9dcda9e8eeb772b6541d0dc94ce91a71044"
+  license "GPL-2.0-only"
+
+  depends_on "qt"
+
+  depends_on xcode: :build
+  on_linux do
```
Comment (carlocab) [inline Formula/texmaker.rb:11]:
```suggestion
  depends_on xcode: :build
  depends_on "qt"

  on_linux do
```

**Comment by albertz** [comment]
Some CI build failed before with this error:
```
 ==> /opt/homebrew/opt/qt/bin/qmake PREFIX=/opt/homebrew/Cellar/texmaker/5.1.2 DESKTOPDIR=/opt/homebrew/Cellar/texmaker/5.1.2/share/applications ICONDIR=/opt/homebrew/Cellar/texmaker/5.1.2/share/pixmaps METAINFODIR=/opt/homebrew/Cellar/texmaker/5.1.2/share/metainfo QMAKE_CXXFLAGS= -config release -spec macx-clang-arm64
  Could not find qmake spec 'macx-clang-arm64'.
  Error processing project file: /private/tmp/texmaker-20211201-63657-132cdp6/texmaker-5.1.2/texmaker.pro
```

That's why I added the Xcode build dependency. Maybe it runs through now. But otherwise, I don't really know how to fix this. I can currently only test it on Linux. I assume sth is wrong with the Qt installation?

File: Formula/texmaker.rb [hunk 3]
```diff
@@ -5,15 +5,14 @@ class Texmaker < Formula
   sha256 "526896f2c1ae561130eec7aae815b9dcda9e8eeb772b6541d0dc94ce91a71044"
   license "GPL-2.0-only"
 
+  depends_on xcode: :build
   depends_on "qt"
 
-  depends_on xcode: :build
   on_linux do
     depends_on "gcc"
+    fails_with gcc: "5"
```
Comment (carlocab) [inline Formula/texmaker.rb:13]: Isn't this failure generic? I don't think you need to keep this in the `on_linux` block.

Comment (albertz) [reply Formula/texmaker.rb:13]: Ok. I am anyway not sure on this. I found some comment in another Qt-based formula which mentioned that GCC 5 causes problems. But doesn't e.g. Mac anyway uses Clang instead?

**Comment by carlocab** [comment]
I think the spec `macx-clang` (without the `-arm64`) is what works for Qt6.

**Comment by albertz** [comment]
Ah, that goes a bit further.

12-arm64:
```
...
  ==> /opt/homebrew/opt/qt/bin/qmake PREFIX=/opt/homebrew/Cellar/texmaker/5.1.2 DESKTOPDIR=/opt/homebrew/Cellar/texmaker/5.1.2/share/applications ICONDIR=/opt/homebrew/Cellar/texmaker/5.1.2/share/pixmaps METAINFODIR=/opt/homebrew/Cellar/texmaker/5.1.2/share/metainfo QMAKE_CXXFLAGS= -config release -spec macx-clang
  Info: creating stash file /private/tmp/texmaker-20211201-20282-68d14a/texmaker-5.1.2/.qmake.stash
  Project WARNING: Qt has only been tested with version 11.0 of the platform SDK, you're using .
  Project WARNING: This is an unsupported configuration. You may experience build issues, and by using
  Project WARNING: the 12.0 SDK you are opting in to new features that Qt has not been prepared for.
  Project WARNING: Please downgrade the SDK you use to build your app to version 11.0, or configure
  Project WARNING: with CONFIG+=sdk_no_version_check when running qmake to silence this warning.
  Project MESSAGE: This project is using private headers and will therefore be tied to this specific Qt module build version.
  Project MESSAGE: Running this project against other versions of the Qt modules may crash at any arbitrary point.
  Project MESSAGE: This is not a bug, but a result of using Qt internals. You have been warned!
  ==> make
...
  ==> make install
...
  Error: Empty installation
  
==> brew audit texmaker --online --new-formula
Error: install failed
```

I don't really understand what goes wrong there.

File: Formula/texmaker.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  depends_on xcode: :build
+  depends_on "qt"
+
+  on_linux do
+    depends_on "gcc"
+  end
+
+  fails_with gcc: "5"
+
+  patch :p0, :DATA
```
Comment (carlocab) [inline Formula/texmaker.rb:17]: Has this patch been upstreamed? We'll probably want to wait for a version where this is included before adding this.

File: Formula/texmaker.rb [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    args << "#{os}-#{compiler}"
+
+    system Formula["qt"].opt_bin/"qmake", *args
+    system "make"
+    system "make", "install"
+  end
+
+  test do
+    assert_predicate bin/"texmaker", :exist?, "texmaker must exist"
+    assert_predicate bin/"texmaker", :executable?, "texmaker must be executable"
```
Comment (carlocab) [inline Formula/texmaker.rb:40]: Can you propose a more substantial test than checking for the existence of certain files, or checking the version or calling for help? These tests rarely ever fail even for broken installations.

>We want tests that don’t require any user input and test the basic functionality of the application. For example `foo build-foo input.foo` is a good test and (despite their widespread use) `foo --version` and `foo --help` are bad tests. 

Ref. https://docs.brew.sh/Formula-Cookbook#add-a-test-to-the-formula

**Review by carlocab — COMMENTED** [review]
I think it's building an app bundle and installing it into `buildpath`.

You probably need to do something like
```ruby
if OS.mac?
  prefix.install buildpath/"texmaker-5.1.2" => "texmaker.app"
  # Install exec scripts into bin here
end
```
The `# Install exec scripts into bin here` comment needs to be replaced with the appropriate commands, of course.

If you just do
```ruby
prefix.install buildpath/"texmaker-5.1.2" if OS.mac?
```
for now we can have a look at the built bottle and see what needs to be done from there.

File: Formula/texmaker.rb (hunk 4 above)
Comment (albertz) [reply Formula/texmaker.rb:17]: There is no public repository. There is just a mail address. I sent a mail with this patch but did not get any response so far. (I wonder a bit, this Qt 5 and Qt 6 docs say that the code should never have worked like this.)

File: Formula/texmaker.rb (hunk 5 above)
Comment (albertz) [reply Formula/texmaker.rb:40]: Unfortunately the binary does not allow for sth like `--help` or `--version`. It will always start the GUI.

Comment (albertz) [reply Formula/texmaker.rb:40]: I could write another patch which patches the `main()` function to add sth like `--version`. But then again, this is probably nicer to have upstream.

**Comment by albertz** [comment]
CI is not running anymore?

File: Formula/texmaker.rb [hunk 6]
```diff
@@ -39,7 +39,8 @@ def install
 
   test do
     if OS.mac?
-      assert_predicate "Texmaker.app", :exist?, "Texmaker.app must exist"
+      assert_predicate prefix/"Texmaker.app", :exist?, "Texmaker.app must exist"
+      assert_predicate prefix/"Texmaker.app/Contents/MacOS/texmaker", :executable?
```
Comment (carlocab) [inline Formula/texmaker.rb:43]: We should create an exec script for this binary.

File: Formula/texmaker.rb (hunk 4 above)
Comment (carlocab) [reply Formula/texmaker.rb:17]: Could you follow it up? That makes it even more important that the patch is upstreamed before this formula is added.

File: Formula/texmaker.rb (hunk 5 above)
Comment (carlocab) [reply Formula/texmaker.rb]: Can we try to run the binary briefly and check that it doesn't crash? Does it generate any logs?

Comment (albertz) [reply Formula/texmaker.rb]: I think it would need an X server to run. Is this ok for such a test? I don't think it creates any logs.

**Comment by albertz** [comment]
The audit is failing (as you mentioned as well):
```
==> brew audit texmaker --online --new-formula
==> FAILED
Full audit texmaker --online --new-formula output
  ==> Installing 'bundler' gem
  texmaker:
    * Formulae should not require patches to build. Patches should be submitted and accepted upstream first.
  Error: 1 problem in 1 formula detected
```
So this is really mandatory?

File: Formula/texmaker.rb (hunk 5 above)
Comment (carlocab) [reply Formula/texmaker.rb]: Does that mean users who install this formula need to install an X server separately to use this?

**Comment by carlocab** [comment]
> So this is really mandatory?

Yes.

File: Formula/texmaker.rb (hunk 5 above)
Comment (albertz) [reply Formula/texmaker.rb]: I don't know. Whatever Qt requires. I think Qt has many possible backends. I'm not sure which backends are installed by Homebrew.

**Comment by albertz** [comment]
> > So this is really mandatory?
> 
> Yes. The PR template actually asks you to run `brew audit --new texmaker` before submission, so you should've seen this earlier.

I did see it, but I also see a lot of usages of `patch` in other formulas, so I thought this must be an error. So why are other usages of `patch` ok then?

**Comment by albertz** [comment]
Nice, all the CI envs seem to build fine. arm64 seems to be quite a bit faster. Ubuntu took over 19 minutes.

File: Formula/texmaker.rb (hunk 4 above)
Comment (albertz) [reply Formula/texmaker.rb:17]: I got an answer now:

> Thanks for the patch. Will be used in the next release.

**Comment by SMillerDev** [comment]
> I did see it, but I also see a lot of usages of patch in other formulas, so I thought this must be an error. So why are other usages of patch ok then?

Because those were added after the software was added to homebrew. They weren't added in a broken state.

File: Formula/texmaker.rb (hunk 4 above)
Comment (albertz) [reply Formula/texmaker.rb:17]: I further asked when the next release will happen. This is the answer:

> I don't think there will be a new release before end of february 2022.

So it means this PR here has to wait until then?

**Comment by albertz** [comment]
Hey, what should we do here? As I understand, we wait for the patch to be applied upstream? But that will not happen before end of February 2022, according to the author.

**Comment by albertz** [comment]
Can we disable the GitHub bot?

Or do we want that this gets closed and then reopened when it is solved upstream?

Or what exactly should we do here now?

**Comment by SMillerDev** [comment]
Well, if nothing happens here it's stale. I'm fine with remaking this PR when the problems are resolved upstream.

**Pull request closed without merging**